src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,394 @@
+    /// This call graph is then returned to the user.
+    /// </summary>
+    public static class BuildCallGraph
+    {
+        public static CallGraph Apply(QsCompilation compilation)
+        {
+            var walker = new BuildGraph();
+            foreach (var ns in compilation.Namespaces)
+            {
+                walker.Namespaces.OnNamespace(ns);
+            }
+            return walker.SharedState.graph;
+        }
+
+        private class BuildGraph : SyntaxTreeTransformation<TransformationState>
+        {
+            public BuildGraph() : base(new TransformationState())
+            {
+                this.Namespaces = new NamespaceTransformation(this);
+                this.Statements = new StatementTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+                this.StatementKinds = new StatementKindTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+                this.Expressions = new ExpressionTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+                this.ExpressionKinds = new ExpressionKindTransformation(this);
+                this.Types = new TypeTransformation<TransformationState>(this, TransformationOptions.Disabled);
+            }
+        }
+
+        private class TransformationState
+        {
+            internal QsSpecialization spec;
+
+            internal bool inCall = false;
+            internal bool hasAdjointDependency = false;
+            internal bool hasControlledDependency = false;
+
+            internal CallGraph graph = new CallGraph();
+        }
+
+        private class NamespaceTransformation : NamespaceTransformation<TransformationState>
+        {
+            public NamespaceTransformation(SyntaxTreeTransformation<TransformationState> parent) : base(parent, TransformationOptions.NoRebuild) { }
+
+            public override QsSpecialization OnSpecializationDeclaration(QsSpecialization spec)
+            {
+                SharedState.spec = spec;
+                return base.OnSpecializationDeclaration(spec);
+            }
+        }
+
+        private class ExpressionKindTransformation : ExpressionKindTransformation<TransformationState>
+        {
+            public ExpressionKindTransformation(SyntaxTreeTransformation<TransformationState> parent) : base(parent, TransformationOptions.NoRebuild) { }
+
+            private ExpressionKind HandleCall(TypedExpression method, TypedExpression arg)
+            {
+                var contextInCall = SharedState.inCall;
+                SharedState.inCall = true;
+                this.Expressions.OnTypedExpression(method);
+                SharedState.inCall = contextInCall;
+                this.Expressions.OnTypedExpression(arg);
+                return ExpressionKind.InvalidExpr;
+            }
+
+            public override ExpressionKind OnOperationCall(TypedExpression method, TypedExpression arg) => HandleCall(method, arg);
```

Comment:
You would also need the same thing for partial applications. It hence makes more sense to just override OnCallLikeExpression instead of OnOperationCall and OnFunctionCall.